app/scripts/controllers/network/network-controller.js

```diff
@@ -509,4 +525,69 @@ export default class NetworkController extends EventEmitter {
     this._provider = provider;
     this._blockTracker = blockTracker;
   }
+
+  /**
+   * Network Configuration management functions
+   */
+
+  /**
+   * Adds a network configuration if the rpcUrl is not already present on an
+   * existing network configuration. Otherwise updates the entry with the matching rpcUrl.
+   *
+   * @param config - The network configuration.
+   * @param config.rpcUrl - The network configuration RPC URL.
+   * @param config.chainId - The chain ID of the network, as per EIP-155.
+   * @param config.ticker - Currency ticker.
+   * @param config.chainName - Personalized network name.
+   * @param config.rpcPrefs - Personalized preferences.
+   * @returns uuid for the added or updated network configuration
+   */
+  upsertNetworkConfiguration({ rpcUrl, chainId, ticker, chainName, rpcPrefs }) {
+    const networkConfigurations = this.networkConfigurations.getState();
+    const newNetworkConfiguration = {
+      rpcUrl,
+      chainId,
+      ticker,
+      chainName,
+      rpcPrefs,
+    };
+
+    for (const [configUUID, networkConfiguration] of Object.entries(
+      networkConfigurations,
+    )) {
+      if (
+        networkConfiguration.rpcUrl?.toLowerCase() === rpcUrl?.toLowerCase()
+      ) {
+        networkConfigurations[configUUID] = newNetworkConfiguration;
+        this.networkConfigurations.updateState(networkConfigurations);
+        return configUUID;
+      }
+    }
+
+    const uuid = random();
+    networkConfigurations[uuid] = newNetworkConfiguration;
+    this.networkConfigurations.updateState(networkConfigurations);
+
+    return uuid;
+  }
+
+  /**
+   * Removes network configuration from state.
+   *
+   * @param uuid - Custom RPC URL.
+   */
+  removeNetworkConfiguration(uuid) {
+    const networkConfigs = this.networkConfigurations.getState();
+    delete networkConfigs[uuid];
+    this.networkConfigurations.updateState(networkConfigs);
+  }
+
+  /**
+   * Getter for the NetworkConfigurations
+   *
+   * @returns {Object<string, NetworkConfiguration>} An object of network configurations.
+   */
+  getNetworkConfigurations() {
```

Comment:
I noticed that this method isn't used as frequently as it could be. For instance, there are places further down below where we just grab `networkConfigurations` out of the network controller state (`const { networkConfigurations } = this.networkController.store.getState()`). To keep the API minimal, I'm curious whether we need this or whether we can go without it?

Reply:
or just `this.networkController.store.getState().networkConfigurations` with no helper method probably makes more sense.

Reply:
done [here](https://github.com/MetaMask/metamask-extension/pull/17421/commits/5f7e68e316a0ccbc5d0e239737f2c44c088a3f01)

Reply:
whoops didn't actually remove the method. That is done [here](https://github.com/MetaMask/metamask-extension/pull/17421/commits/64fc0d8c7dfc0f905e34e6c79662f4a6707bd067)